pkg/controller/machinepool/awsactuator.go

```diff
@@ -87,6 +92,152 @@ func NewAWSActuator(
+	}
+
+	if len(computePool.Platform.AWS.Zones) == 0 {
+		zones, err := a.fetchAvailabilityZones()
+		if err != nil {
+			return nil, false, errors.Wrap(err, "compute pool not providing list of zones and failed to fetch list of zones")
+		}
+		if len(zones) == 0 {
+			return nil, false, fmt.Errorf("zero zones returned for region %s", cd.Spec.Platform.AWS.Region)
+		}
+		computePool.Platform.AWS.Zones = zones
+	}
+
+	subnets := map[string]string{}
+	// Fetching private subnets from the machinepool and then mapping availability zones to subnets
+	if len(pool.Spec.Platform.AWS.Subnets) > 0 {
+		subnetsByAvailabilityZone, err := a.getPrivateSubnetsByAvailabilityZone(pool)
+		if err != nil {
+			return nil, false, errors.Wrap(err, "describing subnets")
+		}
+		subnets = subnetsByAvailabilityZone
+	}
+
+	machineTemplates := make([]*capiaws.AWSMachineTemplate, len(computePool.Platform.AWS.Zones))
+
+	for i, zone := range computePool.Platform.AWS.Zones {
+		subnet, ok := subnets[zone]
+		if len(subnets) > 0 && !ok {
+			return machineTemplates, false, errors.Errorf("no subnet for zone %s", zone)
+		}
+		subnetRef := &capiaws.AWSResourceReference{}
+		if subnet == "" {
+			subnetRef.Filters = []capiaws.Filter{{
+				Name:   "tag:Name",
+				Values: []string{fmt.Sprintf("%s-private-%s", cd.Spec.ClusterMetadata.InfraID, zone)},
+			}}
+		} else {
+			existingSubnet, err := a.getSubnet(subnet)
+			if err != nil {
+				return machineTemplates, false, errors.Wrap(err, "unable to get subnet")
+			}
+			subnetRef.ID = aws.String(*existingSubnet.SubnetId)
+		}
+		machineTemplate, _ := AWSMachineTemplate(cd.Spec.ClusterMetadata.InfraID, a.amiID, pool, subnetRef, cd.Spec.MachineManagement.TargetNamespace, zone)
+		machineTemplates[i] = machineTemplate
+	}
+	logger.Infof("generated %v machinetemplates", len(machineTemplates))
+	return machineTemplates, true, nil
+}
+
+// GenerateCAPIMachineSets takes a clusterDeployment and returns CAPI MachineSets
+func (a *AWSActuator) GenerateCAPIMachineSets(cd *hivev1.ClusterDeployment, pool *hivev1.MachinePool, logger log.FieldLogger) ([]*capiv1.MachineSet, bool, error) {
+	computePool := baseMachinePool(pool)
+	computePool.Platform.AWS = &installertypesaws.MachinePool{
+		AMIID:        a.amiID,
+		InstanceType: pool.Spec.Platform.AWS.InstanceType,
+		EC2RootVolume: installertypesaws.EC2RootVolume{
+			IOPS: pool.Spec.Platform.AWS.EC2RootVolume.IOPS,
+			Size: pool.Spec.Platform.AWS.EC2RootVolume.Size,
+			Type: pool.Spec.Platform.AWS.EC2RootVolume.Type,
+		},
+		Zones: pool.Spec.Platform.AWS.Zones,
+	}
+
+	if len(computePool.Platform.AWS.Zones) == 0 {
+		zones, err := a.fetchAvailabilityZones()
+		if err != nil {
+			return nil, false, errors.Wrap(err, "compute pool not providing list of zones and failed to fetch list of zones")
+		}
+		if len(zones) == 0 {
+			return nil, false, fmt.Errorf("zero zones returned for region %s", cd.Spec.Platform.AWS.Region)
+		}
+		computePool.Platform.AWS.Zones = zones
+	}
```

Comment:
we seem to be doing this exactly same initialization in the GenerateMachineTemplates , could we create initialize aws stuff function?